src/thunderbots/software/ai/hl/stp/tactic/goalie_tactic.cpp

```diff
@@ -0,0 +1,125 @@
+        // posts
+        Point p1                        = Point(field.friendlyGoalpostNeg().x(),
+                         field.friendlyGoalpostNeg().y() + ROBOT_MAX_RADIUS_METERS);
+        Point p2                        = Point(field.friendlyGoalpostPos().x(),
+                         field.friendlyGoalpostPos().y() - ROBOT_MAX_RADIUS_METERS);
+        Segment goalie_movement_segment = Segment(p1, p2);
+        Segment full_goal_segment =
+            Segment(field.friendlyGoalpostNeg(), field.friendlyGoalpostPos());
+
+        // Our default behavior is to track the ball along the goal line, and always face
+        // the ball
+        Point goalie_pos = closestPointOnSeg(ball.position(), goalie_movement_segment);
+        Angle goalie_orientation = (ball.position() - goalie_pos).orientation();
+        std::unique_ptr<Intent> next_intent;
+
+        auto [intersection1, intersection2] =
+            raySegmentIntersection(ball_ray, full_goal_segment);
+        if (intersection1 && ball.velocity().len() > 0.2)
+        {
+            // The ball is heading towards the net. Move to block the shot
+            goalie_pos  = closestPointOnSeg(*intersection1, goalie_movement_segment);
+            next_intent = move_action.updateStateAndGetNextIntent(
+                *robot, goalie_pos, goalie_orientation, 0.0, false, true);
+            std::cout << "shot" << std::endl;
+        }
+        else if (ball.velocity().len() < 0.2 &&
+                 field.pointInFriendlyDefenseArea(ball.position()))
+        {
+            // If the ball is slow or stationary inside our defense area, chip it out
+            next_intent = chip_action.updateStateAndGetNextIntent(
+                *robot, ball, ball.position(),
+                (ball.position() - field.friendlyGoal()).orientation(), 2);
+            std::cout << "clear ball" << std::endl;
+        }
+        else if (enemy_threat)
+        {
+            // If we have been provided with enemy threat information, use it to
+            // preemptively block their best shot. Calculate the best shot the enemy could
+            // take, ignoring the goalie as an obstacle
+            auto best_shot = Evaluation::calcBestShotOnFriendlyGoal(
+                field, friendly_team, enemy_team, ball.position(),
+                ROBOT_MAX_RADIUS_METERS, {*robot});
+            if (best_shot)
+            {
+                Vector shot_vector = (best_shot->first - enemy_threat->robot.position());
+                Ray shot_ray       = Ray(enemy_threat->robot.position(), shot_vector);
+                auto [best_shot_intersection_1, best_shot_intersection_2] =
+                    raySegmentIntersection(shot_ray, full_goal_segment);
+
+                if (best_shot_intersection_1)
+                {
+                    // Move to block the estimated best shot the enemy could take
+                    goalie_pos = closestPointOnSeg(*best_shot_intersection_1,
```

Comment:
An improvement that could be made here:

Pull the goalie out of the goalie_movement_segment to block more of the shot (if safe).

As I found out last year this is particularly important with shots near our endzone.

Reply:
How would we determine if this is safe? I do agree, and will probably enhance goalie behaviour again in a separate PR.